---
title: "Smoller Brainspan Part 2: with RUV"
output:
  html_document:
    theme: cosmo
    toc: true
    toc_depth: 4
    fig_caption: true
    fig_width: 8
    fig_height: 6
author: "Meeta Mistry"
---

```{r setup, echo=FALSE}

# Setup report details
clientname="Erin Dunn"
clientemail="erindunn@pngu.mgh.harvard.edu"
lablocation="MGH"
analystname="Meeta Mistry"
analystemail="mmistry@hsph.harvard.edu"
```

Array analysis for `r clientname` (`r clientemail`) at `r lablocation`. Contact `r analystname` (`r analystemail`) for additional details. Request from client was:

Test for differential gene expression between brain samples taken at different ages as part of the [BrainSpan](http://www.brainspan.org/) project. Rather than downloading directly from the website, most up to date data can be downloaded from GEO [GSE25219](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE25219). Most recently downloaded from GEO on 10/09/2014.

> RNA from 269 human prefrontal cortex samples ranging from fetal development (negative ages) through aging (80 years) were analyzed on custom 2-color microarrays from the National Human Genome Research Institute (NHGRI) microarray core facility using a reference RNA comprised of a pool of all samples.

## Workflow
Starting from the pre-'cleaned' data from GEO, the dataset contains samples from 16 different brain regions. We focused on the orbitofrontal cortex samples for the remaining analyses in this report. Begin by removing the effects of confounders applying various methods which include isva best described in [Teschendorff A.E. et al., 2011](http://bioinformatics.oxfordjournals.org/content/27/11/1496.long) and RUV-2 (remove unwanted variation) a method described in more detail [Gagnon-Bartsch J.A., 2011](http://biostatistics.oxfordjournals.org/content/13/3/539.short). ISVA is not inlcuded in the final html report but code remains in the markdown file.

## Setup

### Bioconductor and R libraries used

```{r libraries, echo=TRUE}
loadlibs <- function(){
library(ggplot2)
library(gtable)
library(scales)
library(RColorBrewer)
library(GEOquery)
library(affy)
library(arrayQualityMetrics)
library(reshape)
library(xtable)
library(isva)
library(limma)
library(Biobase)
library(gridExtra)
library(CHBUtils)
library(png)
library(stringr)
library(dplyr)
library(ruv)
}
suppressPackageStartupMessages(loadlibs())
```

### Set variables
```{r variables, echo=TRUE}
# Setup directory variables
baseDir <- '.'
dataDir <- file.path(baseDir, "data")
metaDir <- file.path(dataDir, "meta")
resultsDir <- file.path(baseDir, "results")
#covarsfilename <- 'covdesc.txt'
```

## Load the expression data

```{r dataimport GEO}

# Load GEO data
gse_gene <- getGEO(filename=file.path(dataDir, 'geo/GSE25219-GPL5175_series_matrix.txt.gz'))
```

## Load the metadata and extract OFC expression data

```{r metadata from file, echo=TRUE, warning=FALSE, message=FALSE}
pheno_gene <- read.delim(file.path(metaDir, 'brainspan_samples_metadata.txt'), row.names=1)
meta_donor <- read.delim(file.path(metaDir, 'brainspan_donor_metadata.txt'), row.names=1)
  
# Subset data by region and hemisphere
meta_ofc <- pheno_gene[which(pheno_gene$BrainRegion == "OFC" & pheno_gene$Hemisphere == "R"),]
meta_ofc$PMI <- as.numeric(as.character(meta_ofc$PMI))
meta_ofc$Stage <- factor(meta_ofc$Stage)

# Add data
data_ofc <- exprs(gse_gene) 
data_ofc <-data_ofc[,which(colnames(data_ofc) %in% rownames(meta_ofc))]
eset.ofc <- new("ExpressionSet", exprs=data_ofc)

# Add metadata
fetal <- rep("NA", nrow(meta_ofc))
stage.num <- as.numeric(as.character(meta_ofc$Stage))
fetal[which(stage.num <= 7)] <- "Fetal"
fetal[which(stage.num > 7)] <- "Postnatal"
fetal <- factor(fetal)
meta_new <-cbind(meta_ofc, fetal)
meta_new$Stage <- stage.num

meta_new <- droplevels(meta_new)
pData(eset.ofc) <- droplevels(meta_new)
fData (eset.ofc) <- fData(gse_gene)
```


## Orbitfrontal Cortex

```{r, echo=FALSE}

# Age distribution
ggplot(meta_new, aes(factor(Stage), fill=fetal)) +
  geom_bar() +
  ggtitle('Orbitofrontal Cortex: Age distribution') + 
  xlab('Developmental Stage') +
  ylab('Number of Samples') +
  theme(axis.text.x = element_text(colour="grey20",size=15,angle=0,hjust=.5,vjust=.5,face="plain"),
        plot.title = element_text(size = rel(2.0)),
        axis.title = element_text(size = rel(1.25)))
```

## Quality Control changes from new and old GEO version
The results below compare the QC for the orbitofrontal cortex data between the old version of the GEO data we downloaded - described in part 1 - with the QC data we have in the current version.  We did this to get a better understanding of how the expression data may have changed from the previous version.  The left hand plot shows shows a false color heatmap of the distances between arrays in the old version. Patterns in this plot indicate clustering of the arrays by fetal and postnatal age with the exception of two samples. The right hand plot shows the same plot using the current version of the data, where we observe the outliers no longer exist. There is definitely some change in the data as samples cluster somewhat differently, although a bit unsure of the source of change. Regardless of this small change we will continue using the current version.

```{r qc-compare, fig.align='center', echo=FALSE, fig.width=11}

img2 <- readPNG("./results/report_OFC_march2014/hm.png")
img3 <- readPNG("./results/report_OFC/hm.png")
grid.arrange(rasterGrob(img2), rasterGrob(img3), ncol=2)
```



```{r isva, fig.align='center', warning=FALSE, message=FALSE, eval=FALSE, echo=FALSE}

## ISVA: Differential expression analysis
# It makes sense to model the effect of confounding factors on the data as statistically independent random variables, as it better reflects the way the confounding noise is generated. This requires them to be uncorrelated with the primary variable of interest in a non-linear fashion, a stronger condition than the _linear_ uncorrelatedness imposed by an SVD. In order to model CFs as statistically independent variables, isva uses independent component analysis (ICA). Since we are using Stage as an ordinal factor (a special case of categorical), it is going to be problematic to have Stages with only one sample. Therefore we will remove samples from Stages 9 and 10. Alot of differentially expressed genes!

# Remove single sample stages
remove <- which(meta_new$Stage == 9 | meta_new$Stage == 10)
meta.isva <- meta_new[-remove,]
data.isva <- data_ofc[,-remove]
  
# Idenitify confounding variables
cf.m <- meta.isva[,c('PMI', 'RIN')]
factor.log <- as.logical(rep("FALSE",2))
diseaseStage <- ordered(meta.isva$Stage, levels=c(2:8, 12:15))

# Run ISVA
isva.res <- DoISVA(data.isva, diseaseStage, cf.m = cf.m, factor.log, th=0.001)
hist(isva.res$spv, main="P-value distribution from isva", xlab="P-value", col="grey", border=F)
```


```{r topPlot, echo=FALSE, fig.align='center', echo=FALSE, eval=FALSE}

### Comparing the top expression changes with age to evaluate effects on PMI 
# The isava  method generates a very large number of significant genes (6118 genes; FDR=0.001). The next step is to evaluate how effectively the isva has removed confounding effects, leaving us with only age-related changes. Below we have taken the top nine genes from the isva result and plotted expression against Age. In the second panel the same genes are plotted to evaluate expression change with PMI. The expression change is almost identical - ordinal can be a special case of continuous, _if_ the categories are equally spaced or perfectly ordered.  But in our case, the categories or stages are not equally spaced (i.e., there's not a one unit or some other consistent unit difference between stage 1 and stage 2) so this is not modeled correctly (some top genes don't even show change).

# Get top genes from sorted p-value list
top <- which(isva.res$rk <= 9)
topSub <- data_ofc[top, ]

# Merge with phenotype information
df <- melt(topSub)
df <- merge(df, meta_ofc, by.x='X2', by.y='row.names')


p1 <- ggplot(df, aes(x=as.numeric(Stage), y=value)) +
  geom_smooth(method=loess) +
  facet_wrap(~X1) +
  theme(axis.title.x = element_blank(),
        plot.margin = unit(c(1, 0, 1, 1), "lines")) + 
  ggtitle('Age') + 
  ylab('Expression values')

p2 <- ggplot(na.omit(df), aes(x=PMI, y=value)) +
  geom_smooth(method=loess) +
  facet_wrap(~X1) +
  theme(axis.title = element_blank(),  
        axis.text.y = element_blank(),
        plot.background = element_blank(),
        axis.ticks.y = element_blank(),
        plot.margin = unit(c(1, 1, 1, 0), "lines")) + 
  ggtitle('Postmortem Interval')

# Set side-by-side
gt1 <- ggplot_gtable(ggplot_build(p1))
gt2 <- ggplot_gtable(ggplot_build(p2))
newWidth = unit.pmax(gt1$widths[2:3], gt2$widths[2:3])

# Set new size
gt1$widths[2:3] = as.list(newWidth)
gt2$widths[2:3] = as.list(newWidth)

# Arrange
grid.arrange(gt1, gt2, ncol=2)

```


## RUV: remove unwanted variation
We next applied RUV to remove unwanted variation in the data - namely to see if we could remove the effects of confounding variables RIN and PMI confounding effects. The strategy with RUV is to use negative control genes and identify surrogate variables based on the expression of those genes. Negative control genes are genes whose expression levels are known a priori to be truly unassociated with the biological factor of interest.  In other words, genes that we know a prior do not change in their expression over the course of development. 

Our original strategy was to use housekeeping genes as provided in [Eisenberg E., and Levanon E.Y.](http://www.ncbi.nlm.nih.gov/pubmed/12850439) as our list of negative control genes.  However, the diganostic plots from an RUV starter anlaysis did not differ between negative control genes and all other genes. This meant one of two things; either the unwanted variation really is extremely correlated with X, or the
negative controls aren't actually negative controls and they're picking up some biology along with the unwanted factors. We assumed it is the latter, and searched for a better set of negative control genes.

Our current/next strategy, was to generate a list of negative control genes by trying to identify a  list of genes that are associated with our unwanted factors (RIN and PMI) but are unaffected by age. Doing so requires using a cohort in which age is not correlated with either RIN or PMI. For this we used only postnatal samples (Stage 8 or higher), and used liner modeling to identify genes differentially expressed with RIN or PMI.

### Using only postnatal samples RIN and PMI show no correlation with age/stage
Below we have plotted Stage on the x-axis and PMI and RIN on the y-axis for each sample in the postnatal cohort. We observe no correlation between the confounding variables and Stage; so we know that any expression changes we do extract from differential expression analysis fpr each factor are _less_likely_to_be_age-related_.

```{r rin-pmi, echo=TRUE, eval=TRUE}

# Keep only samples >= Stage 8
samples <- row.names(pData(eset.ofc))[which(pData(eset.ofc)$Stage >= 8)]
eset.nc <- eset.ofc[,samples]

# Plot relationship with Age
p1 <- ggplot(pData(eset.nc), aes(x=Stage, y=PMI)) + 
  geom_point(shape=1)  +
  geom_smooth(method=lm) +  # Add linear regression line
  ggtitle('PMI') +
  theme(plot.title = element_text(size = rel(2.0)),
        axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25)))

p2 <- ggplot(pData(eset.nc), aes(x=Stage, y=RIN)) + 
  geom_point(shape=1)  +
  geom_smooth(method=lm) +  # Add linear regression line
  ggtitle('RIN') +
  theme(plot.title = element_text(size = rel(2.0)),
        axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25)))

grid.arrange(p1, p2, ncol=2)
```

### Look for genes associated with these unwanted factors
Linear modeling was performed for each factor independently to identify expression changes. P-values were extracted and distributions are plotted below. The uniform distribution indicates that no genes are significantly changing expression with PMI.

ECD:  Here we then plotted the p-value for the test of association between PMI and gene expression level for all genes.  The take-away from this plot is that there does not appear to be a set of possible negative controls here, as the p-value distribution is uniform suggesting that no genes are particularly sensitive to PMI.

MEETA:  In thinking more about this, I'm not sure we want to look at the p-value distribution.  Wouldn't a  plot of the correlation levels between PMI and expression be more informative?  Then you'd have a list of possible genes whose expression appears to be most highly correlated with PMI.  

The results for RIN, however, suggest that there are some genes that are more sensitive to RIN.  We focused the remaining analyses on 50 genes that appear most sensitive to RIN; these genes served as our negative controls.

MEETA:  Don't we actually want to take the 50 genes that are most INSENSITIVE to RIN as our negative controls?

```{r negctl}
# model PMI
mod <- model.matrix(~PMI, pData(eset.nc))
fit<-lmFit(eset.nc, mod)
fit<-eBayes(fit)

topPMI<-topTable(fit,coef=2,number=nrow(exprs(eset.ofc)), adjust.method="BH")
hist(topPMI$P.Value, col="grey", border=F, main="P-value distribution: PMI", xlab="P-value")

# model RIN
mod <- model.matrix(~RIN, pData(eset.nc))
fit<-lmFit(eset.nc, mod)
fit<-eBayes(fit)

topRIN<-topTable(fit,coef=2,number=nrow(exprs(eset.ofc)), adjust.method="BH")
hist(topRIN$P.Value, col="grey", border=F, main="P-value distribution: RIN", xlab="P-value")
topranked <- which(row.names(exprs(eset.ofc)) %in% row.names(topRIN)[1:50])
```

### Check if genes contribute to fetal vs postnatal differences
ECD: Next, we used 50 of the genes identified in the previous stage.  We then plotted those to see how they were associated with the sample age (defined as fetal versus postnatal).  PC 1 refers to XXX.  PC 2 refers to XXX.


```{r pcacheck}
# Check with PCA
myPca <- prcomp(t(exprs(eset.ofc)[topranked,]))

# Plot first factor of interest
tmpPCAData <- as.data.frame(myPca$x[,1:5])
plot(PC2 ~ PC1, data=tmpPCAData, col=c("black", "red")[pData(eset.ofc)[,"fetal"]], 
     pch=19, main="Negative control genes")
legend("top", inset=.05, title="", legend=levels(pData(eset.ofc)[,"fetal"]), fill=c('black', 'red'), horiz=FALSE)
```

### RUV
The two steps of RUV-2 are: 1) perform factor analysis on the control genes to infer the unwanted factors, and 2) perform a simple linear regression of the observed expression levels on the factor of interest. Since we don't really have a healthy mix of PMI and RIN among the patients within each stage, try RUV with just a two group comparison (fetal versus postnatal).  

```{r setup-for-RUV, echo=TRUE}

# Get X and Y
mod <- model.matrix(~fetal, pData(eset.ofc))
X <- as.matrix(mod[,2])
Y <- t(exprs(eset.ofc))

# Assign negative control genes
ctl<-rep("FALSE", nrow(exprs(eset.ofc)))
ctl[topranked]<-"TRUE"
ctl<-as.logical(ctl)
 
# Extract gene information into columns
geneCol <- as.character(fData(eset.ofc)[,'gene_assignment'])
geneNames <- sapply(geneCol, function(x){strsplit(as.character(x), "//")[[1]][2]})
geneNames <- str_trim(geneNames)
geneDesc <- sapply(geneCol, function(x){strsplit(as.character(x), "//")[[1]][3]})
geneChr <- sapply(geneCol, function(x){strsplit(as.character(x), "//")[[1]][4]})
geneChr <- str_trim(geneChr)
geneinfo <- data.frame(geneNames, geneDesc, geneChr, row.names=row.names(fData(eset.ofc)))
```

```{r ruvstarter, eval=FALSE}
# A quick first look at the data
ruv_starter_analysis(Y, X, ctl, geneinfo = t(geneinfo))

# Retry and use kset to indicate specific K
ruv_starter_analysis(Y, X, ctl, geneinfo = t(geneinfo), kset=c(10:15), do_ruv4 = F, do_ruvinv = F, do_ruvrinv = F)
```

Based on the results from the starter analysis we can evaluate the residual RLE plots and p-value distributions to determine an optimal K. We will reference a few key figures below but the full report can be found [here](./ruv/index.html). 

Looking at the residual RLE plots, we see that by the time K = 7 a large amount of unwanted variation has been removed from the negative control genes. The residual RLE plot for K = 10 also looks good, and although only slightly better, the p-value plots help in deciding which to use. By the time K = 15 we see signs of overfitting. Postive control genes will help better determine an optimal K.

The histogram of p-values computed without any adjustment for unwanted variation is far from ideal. There is a large spike near zero, suggesting almost all genes are diferentially expressed. The situation is improved as we increase to K = 10. By the time K = 20 the distribution of the large majority of the genes looks fairly uniform, which is also not ideal. At a K=10 both RLE plots and p-value distribution support a close to optimal adjustment.


```{r ruv-2, echo=FALSE, eval=FALSE}

# RUV-2
ruv.10 <- RUV2(Y, X, ctl, k=10, Z = 1, v = NULL, fullW = NULL, inputcheck = TRUE)

# Histograms
hist (ruv.10$p, col="grey", border=F, xlab="P-value", main="RUV2 with Stage as continuous factor")

# Get SVA object
svaobj<-as.matrix(ruvfit$W)

# Use cleaning function
regressClean<-function(y,mod, svaobj,P=ncol(mod)) {
  X=cbind(mod,svaobj) 
  Hat=solve(t(X)%*%X)%*%t(X) 
  beta=(Hat%*%t(y))
  cleany=y-t(as.matrix(X[,-c(1:P)])%*%beta[-c(1:P),])
  return(cleany)
}

```




